File: codes/core-service/app/routes/plans.py
import uuid

from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.deps import get_db
from app.models.user import User
from app.schemas.plan import (
    PlanApplyRequest,
    PlanApplyRunOut,
    PlanCommentCreate,
    PlanCommentOut,
    PlanCommentUpdate,
    PlanFavoriteOut,
    PlanFavoriteUpdate,
    PlanConfirmRunOut,
    PlanSlotOut,
    PlanSlotPut,
    PlanSubscribeOut,
    PlanSubscribeRequest,
    PlanTemplateCreate,
    PlanTemplateOut,
    PlanTemplateUpdate,
)
from app.services.plan_api import (
    build_comment_out,
    build_slot_out,
    build_template_out,
    create_plan_comment,
    create_plan_template,
    delete_plan_comment,
    delete_plan_template,
    list_plan_comments,
    mark_notification_read,
    replace_slots,
    update_plan_comment,
    update_plan_favorite,
    update_plan_template,
)
from app.services.plan_apply import (
    apply_plan_mode,
    build_apply_run_out,
    cancel_subscription,
    confirm_apply_run,
    skip_apply_run,
    subscribe_plan,
)

router = APIRouter(prefix="/plan-templates", tags=["plan-templates"])
subscription_router = APIRouter(prefix="/plan-subscriptions", tags=["plan-subscriptions"])
apply_run_router = APIRouter(prefix="/plan-apply-runs", tags=["plan-apply-runs"])
notification_router = APIRouter(prefix="/plan-notifications", tags=["plan-notifications"])


@router.post("", response_model=PlanTemplateOut, status_code=status.HTTP_201_CREATED)
def create_template(
    payload: PlanTemplateCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    template = create_plan_template(db, user, payload)
    return build_template_out(db, template)


@router.patch("/{template_id}", response_model=PlanTemplateOut)
def patch_template(
    template_id: uuid.UUID,
    payload: PlanTemplateUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    template = update_plan_template(db, user, template_id, payload)
    return build_template_out(db, template)


@router.delete("/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_template(
    template_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    delete_plan_template(db, user, template_id)


@router.put("/{template_id}/slots", response_model=list[PlanSlotOut])
def put_template_slots(
    template_id: uuid.UUID,
    slots: list[PlanSlotPut],
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    rows = replace_slots(db, user, template_id, slots)
    return [build_slot_out(row) for row in rows]


@router.patch("/{template_id}/favorite", response_model=PlanFavoriteOut)
def patch_template_favorite(
    template_id: uuid.UUID,
    payload: PlanFavoriteUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return update_plan_favorite(db, user, template_id, payload)


@router.post(
    "/{template_id}/apply",
    response_model=PlanApplyRunOut,
    status_code=status.HTTP_201_CREATED,
)
def apply_template(
    template_id: uuid.UUID,
    payload: PlanApplyRequest,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    run = apply_plan_mode(
        db,
        user,
        template_id,
        payload.workspace_id,
        payload.project_id,
        payload.period_start,
    )
    return build_apply_run_out(db, run)


@router.post(
    "/{template_id}/subscribe",
    response_model=PlanSubscribeOut,
    status_code=status.HTTP_201_CREATED,
)
def subscribe_template(
    template_id: uuid.UUID,
    payload: PlanSubscribeRequest,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return subscribe_plan(
        db,
        user,
        template_id,
        payload.workspace_id,
        payload.project_id,
        payload.timezone,
    )


@router.get("/{template_id}/comments", response_model=list[PlanCommentOut])
def get_template_comments(
    template_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    rows = list_plan_comments(db, user, template_id)
    return [build_comment_out(db, row) for row in rows]


@router.post(
    "/{template_id}/comments",
    response_model=PlanCommentOut,
    status_code=status.HTTP_201_CREATED,
)
def add_template_comment(
    template_id: uuid.UUID,
    payload: PlanCommentCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    comment = create_plan_comment(db, user, template_id, payload)
    return build_comment_out(db, comment)


@router.patch("/{template_id}/comments/{comment_id}", response_model=PlanCommentOut)
def patch_template_comment(
    template_id: uuid.UUID,
    comment_id: uuid.UUID,
    payload: PlanCommentUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    comment = update_plan_comment(db, user, template_id, comment_id, payload)
    return build_comment_out(db, comment)


@router.delete(
    "/{template_id}/comments/{comment_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def remove_template_comment(
    template_id: uuid.UUID,
    comment_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    delete_plan_comment(db, user, template_id, comment_id)


@subscription_router.post("/{subscription_id}/cancel", status_code=status.HTTP_204_NO_CONTENT)
def cancel_plan_subscription(
    subscription_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    cancel_subscription(db, user, subscription_id)


@apply_run_router.post("/{run_id}/confirm", response_model=PlanConfirmRunOut)
def confirm_plan_apply_run(
    run_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    run, template_updated = confirm_apply_run(db, user, run_id)
    base = build_apply_run_out(db, run)
    return PlanConfirmRunOut(**base.model_dump(), template_updated=template_updated)


@apply_run_router.post("/{run_id}/skip", response_model=PlanApplyRunOut)
def skip_plan_apply_run(
    run_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    run = skip_apply_run(db, user, run_id)
    return build_apply_run_out(db, run)


@notification_router.post("/{notification_id}/read", status_code=status.HTTP_204_NO_CONTENT)
def read_plan_notification(
    notification_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    mark_notification_read(db, user, notification_id)

File: codes/web/src/lib/api/plans.ts
import { apiFetch } from "@/lib/api";
import type { components } from "@/types/api/generated";
import type { PlanFilterValues } from "@/components/plans/PlanFilters";

export type PlanTemplateCreate = components["schemas"]["PlanTemplateCreate"];
export type PlanTemplateUpdate = components["schemas"]["PlanTemplateUpdate"];
export type PlanTemplateOut = components["schemas"]["PlanTemplateOut"];
export type PlanSlotPut = components["schemas"]["PlanSlotPut"];
export type PlanSlotOut = components["schemas"]["PlanSlotOut"];
export type PlanApplyRequest = components["schemas"]["PlanApplyRequest"];
export type PlanApplyRunOut = components["schemas"]["PlanApplyRunOut"];
export type PlanSubscribeRequest = components["schemas"]["PlanSubscribeRequest"];
export type PlanSubscribeOut = components["schemas"]["PlanSubscribeOut"];
export type PlanConfirmRunOut = components["schemas"]["PlanConfirmRunOut"];
export type PlanCommentCreate = components["schemas"]["PlanCommentCreate"];
export type PlanCommentOut = components["schemas"]["PlanCommentOut"];
export type PlanListOut = components["schemas"]["PlanListOut"];
export type PlanCardOut = components["schemas"]["PlanCardOut"];
export type PlanDetailOut = components["schemas"]["PlanDetailOut"];
export type PlanImportedListOut = components["schemas"]["PlanImportedListOut"];
export type PlanImportedRowOut = components["schemas"]["PlanImportedRowOut"];
export type PlanImportedRunOut = components["schemas"]["PlanImportedRunOut"];
export type PlanRunItemOut = components["schemas"]["PlanRunItemOut"];
export type PlanSubscribedListOut = components["schemas"]["PlanSubscribedListOut"];
export type PlanSubscribedRowOut = components["schemas"]["PlanSubscribedRowOut"];
export type PlanPendingRunOut = components["schemas"]["PlanPendingRunOut"];
export type PlanNotificationListOut = components["schemas"]["PlanNotificationListOut"];
export type PlanNotificationOut = components["schemas"]["PlanNotificationOut"];
export type PlanFavoriteOut = { template_id: string; is_favorite: boolean };

export type FetchPlanListParams = {
  tab?: string;
  q?: string;
  visibility?: string;
  creator_q?: string;
  tag?: string[];
  period_kind?: string;
  usage_kind?: string;
  favorite?: boolean;
  limit?: number;
  offset?: number;
};

export type PlanPageParams = {
  q?: string;
  creator_q?: string;
  tag?: string[];
  period_kind?: string;
  usage_kind?: string;
  favorite?: boolean;
  limit?: number;
  offset?: number;
};

export function planFilterQueryParams(filters: PlanFilterValues): Omit<PlanPageParams, "limit" | "offset"> {
  return {
    q: filters.q.trim() || undefined,
    creator_q: filters.creator_q.trim() || undefined,
    period_kind: filters.period_kind || undefined,
    usage_kind: filters.usage_kind || undefined,
    tag: filters.tags.length > 0 ? filters.tags : undefined,
    favorite:
      filters.favorite === "true" ? true : filters.favorite === "false" ? false : undefined,
  };
}

function withQuery(
  path: string,
  params: Record<string, string | number | boolean | undefined | string[]>,
): string {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === null || value === "") continue;
    if (Array.isArray(value)) {
      for (const item of value) {
        if (item !== undefined && item !== null && item !== "") search.append(key, item);
      }
      continue;
    }
    search.set(key, String(value));
  }
  const qs = search.toString();
  return qs ? `${path}?${qs}` : path;
}

export function fetchPlanList(
  token: string,
  params: FetchPlanListParams = {},
): Promise<PlanListOut> {
  return apiFetch<PlanListOut>(
    withQuery("/views/plans", {
      tab: params.tab,
      q: params.q,
      visibility: params.visibility,
      creator_q: params.creator_q,
      tag: params.tag,
      period_kind: params.period_kind,
      usage_kind: params.usage_kind,
      favorite: params.favorite,
      limit: params.limit,
      offset: params.offset,
    }),
    { token },
  );
}

export function fetchPlanDetail(token: string, planId: string): Promise<PlanDetailOut> {
  return apiFetch<PlanDetailOut>(`/views/plans/${planId}`, { token });
}

export function fetchImportedPlans(
  token: string,
  opts: PlanPageParams = {},
): Promise<PlanImportedListOut> {
  return apiFetch<PlanImportedListOut>(
    withQuery("/views/plans/imported", {
      q: opts.q,
      creator_q: opts.creator_q,
      tag: opts.tag,
      period_kind: opts.period_kind,
      usage_kind: opts.usage_kind,
      favorite: opts.favorite,
      limit: opts.limit,
      offset: opts.offset,
    }),
    { token },
  );
}

export function fetchSubscribedPlans(
  token: string,
  opts: PlanPageParams = {},
): Promise<PlanSubscribedListOut> {
  return apiFetch<PlanSubscribedListOut>(
    withQuery("/views/plans/subscribed", {
      q: opts.q,
      creator_q: opts.creator_q,
      tag: opts.tag,
      period_kind: opts.period_kind,
      usage_kind: opts.usage_kind,
      favorite: opts.favorite,
      limit: opts.limit,
      offset: opts.offset,
    }),
    { token },
  );
}

export function fetchPlanNotifications(
  token: string,
  opts: PlanPageParams = {},
): Promise<PlanNotificationListOut> {
  return apiFetch<PlanNotificationListOut>(
    withQuery("/views/plan-notifications", { limit: opts.limit, offset: opts.offset }),
    { token },
  );
}

export function createPlanTemplate(
  token: string,
  payload: PlanTemplateCreate,
): Promise<PlanTemplateOut> {
  return apiFetch<PlanTemplateOut>("/plan-templates", {
    method: "POST",
    token,
    body: JSON.stringify(payload),
  });
}

export function updatePlanTemplate(
  token: string,
  templateId: string,
  payload: PlanTemplateUpdate,
): Promise<PlanTemplateOut> {
  return apiFetch<PlanTemplateOut>(`/plan-templates/${templateId}`, {
    method: "PATCH",
    token,
    body: JSON.stringify(payload),
  });
}

export async function deletePlanTemplate(token: string, templateId: string): Promise<void> {
  await apiFetch<void>(`/plan-templates/${templateId}`, { method: "DELETE", token });
}

export function putPlanSlots(
  token: string,
  templateId: string,
  slots: PlanSlotPut[],
): Promise<PlanSlotOut[]> {
  return apiFetch<PlanSlotOut[]>(`/plan-templates/${templateId}/slots`, {
    method: "PUT",
    token,
    body: JSON.stringify(slots),
  });
}

export function applyPlan(
  token: string,
  templateId: string,
  payload: PlanApplyRequest,
): Promise<PlanApplyRunOut> {
  return apiFetch<PlanApplyRunOut>(`/plan-templates/${templateId}/apply`, {
    method: "POST",
    token,
    body: JSON.stringify(payload),
  });
}

export function subscribePlan(
  token: string,
  templateId: string,
  payload: PlanSubscribeRequest,
): Promise<PlanSubscribeOut> {
  return apiFetch<PlanSubscribeOut>(`/plan-templates/${templateId}/subscribe`, {
    method: "POST",
    token,
    body: JSON.stringify(payload),
  });
}

export async function cancelPlanSubscription(token: string, subscriptionId: string): Promise<void> {
  await apiFetch<void>(`/plan-subscriptions/${subscriptionId}/cancel`, {
    method: "POST",
    token,
  });
}

export function confirmPlanApplyRun(token: string, runId: string): Promise<PlanConfirmRunOut> {
  return apiFetch<PlanConfirmRunOut>(`/plan-apply-runs/${runId}/confirm`, {
    method: "POST",
    token,
  });
}

export function skipPlanApplyRun(token: string, runId: string): Promise<PlanApplyRunOut> {
  return apiFetch<PlanApplyRunOut>(`/plan-apply-runs/${runId}/skip`, {
    method: "POST",
    token,
  });
}

export function listPlanComments(token: string, templateId: string): Promise<PlanCommentOut[]> {
  return apiFetch<PlanCommentOut[]>(`/plan-templates/${templateId}/comments`, { token });
}

export function addPlanComment(
  token: string,
  templateId: string,
  payload: PlanCommentCreate,
): Promise<PlanCommentOut> {
  return apiFetch<PlanCommentOut>(`/plan-templates/${templateId}/comments`, {
    method: "POST",
    token,
    body: JSON.stringify(payload),
  });
}

export async function markPlanNotificationRead(
  token: string,
  notificationId: string,
): Promise<void> {
  await apiFetch<void>(`/plan-notifications/${notificationId}/read`, {
    method: "POST",
    token,
  });
}

export function updatePlanFavorite(
  token: string,
  templateId: string,
  isFavorite: boolean,
): Promise<PlanFavoriteOut> {
  return apiFetch<PlanFavoriteOut>(`/plan-templates/${templateId}/favorite`, {
    method: "PATCH",
    token,
    body: JSON.stringify({ is_favorite: isFavorite }),
  });
}
